Database cleanup errors

I am receiving this error in my logs on each scan.

I have tried deleting the db and starting from scratch but still the same error. No idea if it matters or what effect it is having.

This is running v2.1.1.

```
photoview_1    | 2021/02/16 21:47:01 /app/scanner/cleanup_media.go:106 Error 1451: Cannot delete or update a parent row: a foreign key constraint fails (`photoview`.`user_albums`, CONSTRAINT `fk_user_albums_album` FOREIGN KEY (`album_id`) REFERENCES `albums` (`id`))
photoview_1    | [25.553ms] [rows:0] DELETE FROM `albums` WHERE id IN (37,88,129,339,524,546,798,1089)
photoview_1    | 2021/02/16 21:47:01 ERROR: Could not delete old albums from database:
photoview_1    | Error 1451: Cannot delete or update a parent row: a foreign key constraint fails (`photoview`.`user_albums`, CONSTRAINT `fk_user_albums_album` FOREIGN KEY (`album_id`) REFERENCES `albums` (`id`))
```

--- a/api/database/database.go
+++ b/api/database/database.go
@@ -103,6 +103,11 @@ func ConfigureDatabase(config *gorm.Config) (*gorm.DB, error) {
 		return nil, err
 	}
 
+	// Manually enable foreign keys for sqlite, as this isn't done by default
+	if drivers.DatabaseDriver() == drivers.DatabaseDriverSqlite {
+		db.Exec("PRAGMA foreign_keys = ON")
+	}
+
 	return db, nil
 }
 

--- a/api/go.mod
+++ b/api/go.mod
@@ -19,6 +19,7 @@ require (
 	github.com/joho/godotenv v1.3.0
 	github.com/mattn/go-sqlite3 v1.14.6 // indirect
 	github.com/mitchellh/mapstructure v1.4.1 // indirect
+	github.com/otiai10/copy v1.6.0 // indirect
 	github.com/pkg/errors v0.9.1
 	github.com/sabhiram/go-gitignore v0.0.0-20201211210132-54b8a0bf510f
 	github.com/stretchr/objx v0.3.0 // indirect

--- a/api/graphql/resolvers/user.go
+++ b/api/graphql/resolvers/user.go
@@ -343,7 +343,7 @@ func (r *mutationResolver) UserRemoveRootAlbum(ctx context.Context, userID int, 
 		}
 
 		// Reload faces as media might have been deleted
-		if face_detection.GlobalFaceDetector == nil {
+		if face_detection.GlobalFaceDetector != nil {
 			if err := face_detection.GlobalFaceDetector.ReloadFacesFromDatabase(r.Database); err != nil {
 				return nil, err
 			}

--- a/api/routes/authenticate_routes_test.go
+++ b/api/routes/authenticate_routes_test.go
@@ -88,8 +88,8 @@ func TestAuthenticateRoute(t *testing.T) {
 
 			assert.NoError(t, err)
 			assert.True(t, success)
-			assert.Equal(t, responseMessage, "success")
-			assert.Equal(t, responseStatus, http.StatusAccepted)
+			assert.Equal(t, "success", responseMessage)
+			assert.Equal(t, http.StatusAccepted, responseStatus)
 		})
 	})
 
@@ -103,8 +103,8 @@ func TestAuthenticateRoute(t *testing.T) {
 
 			assert.NoError(t, err)
 			assert.True(t, success)
-			assert.Equal(t, responseMessage, "success")
-			assert.Equal(t, responseStatus, http.StatusAccepted)
+			assert.Equal(t, "success", responseMessage)
+			assert.Equal(t, http.StatusAccepted, responseStatus)
 		})
 
 		t.Run("Request without access token", func(t *testing.T) {
@@ -114,8 +114,8 @@ func TestAuthenticateRoute(t *testing.T) {
 
 			assert.Error(t, err)
 			assert.False(t, success)
-			assert.Equal(t, responseMessage, "unauthorized")
-			assert.Equal(t, responseStatus, http.StatusForbidden)
+			assert.Equal(t, "unauthorized", responseMessage)
+			assert.Equal(t, http.StatusForbidden, responseStatus)
 		})
 
 		expire := time.Now().Add(time.Hour * 24 * 30)
@@ -139,8 +139,8 @@ func TestAuthenticateRoute(t *testing.T) {
 
 			assert.NoError(t, err)
 			assert.True(t, success)
-			assert.Equal(t, responseMessage, "success")
-			assert.Equal(t, responseStatus, http.StatusAccepted)
+			assert.Equal(t, "success", responseMessage)
+			assert.Equal(t, http.StatusAccepted, responseStatus)
 		})
 	})
 

--- a/api/scanner/cleanup_media.go
+++ b/api/scanner/cleanup_media.go
@@ -83,7 +83,6 @@ func deleteOldUserAlbums(db *gorm.DB, scannedAlbums []*models.Album, user *model
 		Table("user_albums").
 		Joins("JOIN albums ON user_albums.album_id = albums.id").
 		Where("user_id = ?", user.ID).
-		// Where("album_id IN (?)", userAlbumIDs).
 		Where("album_id NOT IN (?)", scannedAlbumIDs)
 
 	if err := query.Find(&deleteAlbums).Error; err != nil {
@@ -109,7 +108,7 @@ func deleteOldUserAlbums(db *gorm.DB, scannedAlbums []*models.Album, user *model
 
 	// Delete old albums from database
 	err := db.Transaction(func(tx *gorm.DB) error {
-		if err := tx.Model(&user).Association("Albums").Delete(deleteAlbums); err != nil {
+		if err := tx.Where("album_id IN ?", deleteAlbumIDs).Delete(&models.UserAlbums{}).Error; err != nil {
 			return err
 		}
 

--- a/api/scanner/cleanup_media.go
+++ b/api/scanner/cleanup_media.go
@@ -108,11 +108,11 @@ func deleteOldUserAlbums(db *gorm.DB, scannedAlbums []*models.Album, user *model
 
 	// Delete old albums from database
 	err := db.Transaction(func(tx *gorm.DB) error {
-		if err := tx.Where("album_id IN ?", deleteAlbumIDs).Delete(&models.UserAlbums{}).Error; err != nil {
+		if err := tx.Where("album_id IN (?)", deleteAlbumIDs).Delete(&models.UserAlbums{}).Error; err != nil {
 			return err
 		}
 
-		if err := tx.Where("id IN ?", deleteAlbumIDs).Delete(models.Album{}).Error; err != nil {
+		if err := tx.Where("id IN (?)", deleteAlbumIDs).Delete(models.Album{}).Error; err != nil {
 			return err
 		}
 

--- a/api/scanner/cleanup_media.go
+++ b/api/scanner/cleanup_media.go
@@ -126,7 +126,7 @@ func deleteOldUserAlbums(db *gorm.DB, scannedAlbums []*models.Album, user *model
 	}
 
 	// Reload faces after deleting albums
-	if face_detection.GlobalFaceDetector == nil {
+	if face_detection.GlobalFaceDetector != nil {
 		if err := face_detection.GlobalFaceDetector.ReloadFacesFromDatabase(db); err != nil {
 			deleteErrors = append(deleteErrors, err)
 		}

--- a/api/scanner/cleanup_media_test.go
+++ b/api/scanner/cleanup_media_test.go
@@ -7,6 +7,7 @@ import (
 
 	"github.com/otiai10/copy"
 	"github.com/photoview/photoview/api/graphql/models"
+	"github.com/photoview/photoview/api/scanner/face_detection"
 	"github.com/photoview/photoview/api/test_utils"
 	"github.com/stretchr/testify/assert"
 )
@@ -14,6 +15,10 @@ import (
 func TestCleanupMedia(t *testing.T) {
 	test_utils.FilesystemTest(t)
 	db := test_utils.DatabaseTest(t)
+
+	if !assert.NoError(t, face_detection.InitializeFaceDetector(db)) {
+		return
+	}
 
 	test_dir := t.TempDir()
 	copy.Copy("./test_data", test_dir)
@@ -82,7 +87,15 @@ func TestCleanupMedia(t *testing.T) {
 		assert.Equal(t, 6, countAllMediaURLs())
 	})
 
-	// t.Run("Modify images", func(t *testing.T) {
+	t.Run("Modify images", func(t *testing.T) {
+		assert.NoError(t, os.Rename(path.Join(test_dir, "buttercup_close_summer_yellow.jpg"), path.Join(test_dir, "yellow-flower.jpg")))
+		test_utils.RunScannerAll(t, db)
+		assert.Equal(t, 3, countAllMedia())
+		assert.Equal(t, 6, countAllMediaURLs())
 
-	// })
+		assert.NoError(t, os.Remove(path.Join(test_dir, "lilac_lilac_bush_lilac.jpg")))
+		test_utils.RunScannerAll(t, db)
+		assert.Equal(t, 2, countAllMedia())
+		assert.Equal(t, 4, countAllMediaURLs())
+	})
 }

--- a/api/scanner/cleanup_media_test.go
+++ b/api/scanner/cleanup_media_test.go
@@ -1,0 +1,81 @@
+package scanner_test
+
+import (
+	"os"
+	"path"
+	"testing"
+
+	"github.com/otiai10/copy"
+	"github.com/photoview/photoview/api/graphql/models"
+	"github.com/photoview/photoview/api/test_utils"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCleanupMedia(t *testing.T) {
+	test_utils.FilesystemTest(t)
+	db := test_utils.DatabaseTest(t)
+
+	test_dir := t.TempDir()
+	copy.Copy("./test_data", test_dir)
+
+	countAllMedia := func() int {
+		var all_media []*models.Media
+		if !assert.NoError(t, db.Find(&all_media).Error) {
+			return -1
+		}
+		return len(all_media)
+	}
+
+	countAllMediaURLs := func() int {
+		var all_media_urls []*models.MediaURL
+		if !assert.NoError(t, db.Find(&all_media_urls).Error) {
+			return -1
+		}
+		return len(all_media_urls)
+	}
+
+	pass := "1234"
+	user1, err := models.RegisterUser(db, "user1", &pass, true)
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	user2, err := models.RegisterUser(db, "user2", &pass, true)
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	root_album := models.Album{
+		Title: "root album",
+		Path:  test_dir,
+	}
+
+	if !assert.NoError(t, db.Save(&root_album).Error) {
+		return
+	}
+
+	err = db.Model(user1).Association("Albums").Append(&root_album)
+	if !assert.NoError(t, err) {
+		return
+	}
+	err = db.Model(user2).Association("Albums").Append(&root_album)
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	test_utils.RunScannerOnUser(t, db, user1)
+	assert.Equal(t, 9, countAllMedia())
+	assert.Equal(t, 18, countAllMediaURLs())
+
+	// move faces directory
+	assert.NoError(t, os.Rename(path.Join(test_dir, "faces"), path.Join(test_dir, "faces_moved")))
+	test_utils.RunScannerAll(t, db)
+	assert.Equal(t, 9, countAllMedia())
+	assert.Equal(t, 18, countAllMediaURLs())
+
+	// remove faces_moved directory
+	assert.NoError(t, os.RemoveAll(path.Join(test_dir, "faces_moved")))
+	test_utils.RunScannerAll(t, db)
+	assert.Equal(t, 3, countAllMedia())
+	assert.Equal(t, 6, countAllMediaURLs())
+}

--- a/api/scanner/cleanup_media_test.go
+++ b/api/scanner/cleanup_media_test.go
@@ -63,19 +63,26 @@ func TestCleanupMedia(t *testing.T) {
 		return
 	}
 
-	test_utils.RunScannerOnUser(t, db, user1)
-	assert.Equal(t, 9, countAllMedia())
-	assert.Equal(t, 18, countAllMediaURLs())
+	t.Run("Modify albums", func(t *testing.T) {
 
-	// move faces directory
-	assert.NoError(t, os.Rename(path.Join(test_dir, "faces"), path.Join(test_dir, "faces_moved")))
-	test_utils.RunScannerAll(t, db)
-	assert.Equal(t, 9, countAllMedia())
-	assert.Equal(t, 18, countAllMediaURLs())
+		test_utils.RunScannerOnUser(t, db, user1)
+		assert.Equal(t, 9, countAllMedia())
+		assert.Equal(t, 18, countAllMediaURLs())
 
-	// remove faces_moved directory
-	assert.NoError(t, os.RemoveAll(path.Join(test_dir, "faces_moved")))
-	test_utils.RunScannerAll(t, db)
-	assert.Equal(t, 3, countAllMedia())
-	assert.Equal(t, 6, countAllMediaURLs())
+		// move faces directory
+		assert.NoError(t, os.Rename(path.Join(test_dir, "faces"), path.Join(test_dir, "faces_moved")))
+		test_utils.RunScannerAll(t, db)
+		assert.Equal(t, 9, countAllMedia())
+		assert.Equal(t, 18, countAllMediaURLs())
+
+		// remove faces_moved directory
+		assert.NoError(t, os.RemoveAll(path.Join(test_dir, "faces_moved")))
+		test_utils.RunScannerAll(t, db)
+		assert.Equal(t, 3, countAllMedia())
+		assert.Equal(t, 6, countAllMediaURLs())
+	})
+
+	// t.Run("Modify images", func(t *testing.T) {
+
+	// })
 }

--- a/api/scanner/exif/exif.go
+++ b/api/scanner/exif/exif.go
@@ -9,19 +9,19 @@ import (
 	"github.com/photoview/photoview/api/graphql/models"
 )
 
-type exifParser interface {
+type ExifParser interface {
 	ParseExif(media_path string) (*models.MediaEXIF, error)
 }
 
-var globalExifParser exifParser
+var globalExifParser ExifParser
 
 func InitializeEXIFParser() {
 	// Decide between internal or external Exif parser
-	exiftoolParser, err := newExiftoolParser()
+	exiftoolParser, err := NewExiftoolParser()
 
 	if err != nil {
 		log.Printf("Failed to get exiftool, using internal exif parser instead: %v\n", err)
-		globalExifParser = &internalExifParser{}
+		globalExifParser = NewInternalExifParser()
 	} else {
 		log.Println("Found exiftool")
 		globalExifParser = exiftoolParser

--- a/api/scanner/exif/exif_parser_external.go
+++ b/api/scanner/exif/exif_parser_external.go
@@ -15,7 +15,7 @@ type externalExifParser struct {
 	dataLoader *dataloader.ExiftoolLoader
 }
 
-func newExiftoolParser() (*externalExifParser, error) {
+func NewExiftoolParser() (ExifParser, error) {
 	et, err := exiftool.NewExiftool(exiftool.NoPrintConversion())
 
 	if err != nil {

--- a/api/scanner/exif/exif_parser_internal.go
+++ b/api/scanner/exif/exif_parser_internal.go
@@ -16,7 +16,11 @@ import (
 // internalExifParser is an exif parser that parses the media without the use of external tools
 type internalExifParser struct{}
 
-func (p *internalExifParser) ParseExif(media_path string) (returnExif *models.MediaEXIF, returnErr error) {
+func NewInternalExifParser() ExifParser {
+	return internalExifParser{}
+}
+
+func (p internalExifParser) ParseExif(media_path string) (returnExif *models.MediaEXIF, returnErr error) {
 	photoFile, err := os.Open(media_path)
 	if err != nil {
 		return nil, err

--- a/api/scanner/exif/exif_test.go
+++ b/api/scanner/exif/exif_test.go
@@ -1,4 +1,4 @@
-package exif
+package exif_test
 
 import (
 	"fmt"
@@ -9,6 +9,7 @@ import (
 
 	"github.com/barasher/go-exiftool"
 	"github.com/photoview/photoview/api/graphql/models"
+	"github.com/photoview/photoview/api/scanner/exif"
 	"github.com/photoview/photoview/api/test_utils"
 	"github.com/stretchr/testify/assert"
 )
@@ -22,18 +23,18 @@ func TestExifParsers(t *testing.T) {
 
 	parsers := []struct {
 		name   string
-		parser exifParser
+		parser exif.ExifParser
 	}{
 		{
 			name:   "internal",
-			parser: &internalExifParser{},
+			parser: exif.NewInternalExifParser(),
 		},
 	}
 
-	if externalParser, err := newExiftoolParser(); err == nil {
+	if externalParser, err := exif.NewExiftoolParser(); err == nil {
 		parsers = append(parsers, struct {
 			name   string
-			parser exifParser
+			parser exif.ExifParser
 		}{
 			name:   "external",
 			parser: externalParser,

--- a/api/scanner/scanner_test.go
+++ b/api/scanner/scanner_test.go
@@ -6,7 +6,6 @@ import (
 	"time"
 
 	"github.com/photoview/photoview/api/graphql/models"
-	"github.com/photoview/photoview/api/scanner"
 	"github.com/photoview/photoview/api/scanner/face_detection"
 	"github.com/photoview/photoview/api/test_utils"
 	"github.com/stretchr/testify/assert"
@@ -40,20 +39,11 @@ func TestFullScan(t *testing.T) {
 		return
 	}
 
-	if !assert.NoError(t, scanner.InitializeScannerQueue(db)) {
-		return
-	}
-
 	if !assert.NoError(t, face_detection.InitializeFaceDetector(db)) {
 		return
 	}
 
-	if !assert.NoError(t, scanner.AddUserToQueue(user)) {
-		return
-	}
-
-	// wait for all jobs to finish
-	scanner.CloseScannerQueue()
+	test_utils.RunScannerOnUser(t, db, user)
 
 	var all_media []*models.Media
 	if !assert.NoError(t, db.Find(&all_media).Error) {

--- a/api/scanner/scanner_user.go
+++ b/api/scanner/scanner_user.go
@@ -23,7 +23,7 @@ func getPhotoviewIgnore(ignorePath string) ([]string, error) {
 	// Open .photoviewignore file, if exists
 	photoviewIgnoreFile, err := os.Open(path.Join(ignorePath, ".photoviewignore"))
 	if err != nil {
-		if err == os.ErrNotExist {
+		if os.IsNotExist(err) {
 			return photoviewIgnore, nil
 		}
 		return photoviewIgnore, err

--- a/api/test_utils/scanner_helpers.go
+++ b/api/test_utils/scanner_helpers.go
@@ -1,0 +1,36 @@
+package test_utils
+
+import (
+	"testing"
+
+	"github.com/photoview/photoview/api/graphql/models"
+	"github.com/photoview/photoview/api/scanner"
+	"github.com/stretchr/testify/assert"
+	"gorm.io/gorm"
+)
+
+func RunScannerOnUser(t *testing.T, db *gorm.DB, user *models.User) {
+	if !assert.NoError(t, scanner.InitializeScannerQueue(db)) {
+		return
+	}
+
+	if !assert.NoError(t, scanner.AddUserToQueue(user)) {
+		return
+	}
+
+	// wait for all jobs to finish
+	scanner.CloseScannerQueue()
+}
+
+func RunScannerAll(t *testing.T, db *gorm.DB) {
+	if !assert.NoError(t, scanner.InitializeScannerQueue(db)) {
+		return
+	}
+
+	if !assert.NoError(t, scanner.AddAllToQueue()) {
+		return
+	}
+
+	// wait for all jobs to finish
+	scanner.CloseScannerQueue()
+}

--- a/api/utils/environment_variables.go
+++ b/api/utils/environment_variables.go
@@ -66,13 +66,13 @@ func (v EnvironmentVariable) GetBool() bool {
 
 // ShouldServeUI whether or not the "serve ui" option is enabled
 func ShouldServeUI() bool {
-	return EnvServeUI.GetValue() == "1"
+	return EnvServeUI.GetBool()
 }
 
 // DevelopmentMode describes whether or not the server is running in development mode,
 // and should thus print debug informations and enable other features related to developing.
 func DevelopmentMode() bool {
-	return EnvDevelopmentMode.GetValue() == "1"
+	return EnvDevelopmentMode.GetBool()
 }
 
 // UIPath returns the value from where the static UI files are located if SERVE_UI=1

